Source/WebCore/page/SecurityOriginData.cpp

```diff
@@ -66,9 +67,23 @@ SecurityOriginData SecurityOriginData::fromFrame(Frame* frame)
     return document->securityOrigin().data();
 }
 
+SecurityOriginData SecurityOriginData::fromURL(const URL& url, ShouldSkipOpaqueSecurityOriginCheck shouldSkipOpaqueSecurityOriginCheck)
+{
+    bool shouldCreateOpaqueOrigin = ShouldSkipOpaqueSecurityOriginCheck::No == shouldSkipOpaqueSecurityOriginCheck && shouldTreatAsOpaqueOrigin(url);
+    ASSERT(shouldCreateOpaqueOrigin || !url.protocol().isNull() || !url.host().isNull());
+
+    if (shouldCreateOpaqueOrigin || (url.protocol().isNull() && url.host().isNull()))
+        return createOpaque();
+    return SecurityOriginData {
```

Comment:
Shouldn't we create an opaque URL if url.protocol() is null/empty and url.host() is null/empty?
Right now, it seems we'll construct a SecurityOriginData with protocol&host being the emptyString, which used to be recognized as opaque, but not anymore after your change?

Reply:
Yes, good catch. I moved the empty/null checks into the constructor. It's not very pretty, so happy to hear suggestions.